### DISCUSSION TITLE
mxmlc / compc and ANEs

### DISCUSSION THREAD

**Discussion by marchbold** [airsdk/Adobe-Runtime-Support#1443] [Ideas]
Is there any reason `mxmlc` and `compc` seem to ignore ANEs added as `external-library-path`?

Currently you have to either extract a swc from the ane (what some IDEs do) or simply rename the `.ane` to `.swc` (normally what we do in our builds). The ANE format at it's base level is the same as a SWC so it's not surprising this works. 

I wonder if it's something we could easily add into the compilers just to accept ANE's as external libraries directly?

**Comment by ajwfrost** [comment]
Hi @marchbold  - I've just come across this topic, and wondered if it's still an issue? I've just run a command line which works:
```
mxmlc -debug AIRInfo.as -compiler.external-library-path+=ane\com.distriqt.Core.ane
```
This was with a 50.2 AIR SDK. But I don't see any changes that we'd made in the compiler.jar to suddenly change the behaviour around this, wondering if there may be a difference in the Flex compiler vs the AIR one perhaps?

thanks

**Reply by marchbold** [reply]
Right, so this was in reference to the ant build scripts using the `flexTasks.jar` and yes the issue still exists. 

eg the following fails to find the `com.distriqt.Core` extension in the `ane` directory: 

```xml
<?xml version="1.0" encoding="UTF-8"?>
<project default="build" name="app">

    <property environment="env."/>
    <fail unless="env.AIR_HOME" message="AIR_HOME needs to be defined as an environment variable or in the Ant build." />
    <property name="AIR_HOME" value="${env.AIR_HOME}" />
    <property name="AIR_SDK_HOME" value="${AIR_HOME}" />
    <taskdef resource="flexTasks.tasks" classpath="${AIR_HOME}/ant/lib/flexTasks.jar" />
    <dirname property="project.basedir" file="${ant.file.app}" />

    <target name="build">
        <mxmlc file="${project.basedir}/src/TestMXMLC.as">
            <load-config filename="${AIR_HOME}/frameworks/air-config.xml"/> 
            <external-library-path dir="${project.basedir}/ane" append="true" >
                <include name="**/*.ane"/> 
            </external-library-path> 
        </mxmlc>
    </target>

</project>
```

You are right that the equivalent command seems to work:

```sh
mxmlc src/TestMXMLC.as -compiler.external-library-path+=ane
```

But it works if I rename all the ane's to have a swc extension and update the command:

```xml
            <external-library-path dir="${project.basedir}/ane" append="true" >
                <include name="**/*.swc"/> 
            </external-library-path> 
```

So something in the `flexTasks.jar` then?

**Reply by marchbold** [reply]
TestMXMLC.as:

```actionscript
package
{
    import flash.display.Sprite;
    import com.distriqt.extension.core.Core;

    public class TestMXMLC extends Sprite
    {
        public function TestMXMLC()
        {
            Core.init();
        }
    }

}
```

**Reply by marchbold** [reply]
Ant output: 

```
Buildfile: /Users/marchbold/work/distriqt/airsdk/tests/mxmlc-anes/build.xml

build:
    [mxmlc] Loading configuration: /Users/marchbold/work/sdks/air/AIRSDK_50.2.3/frameworks/air-config.xml
    [mxmlc] 
    [mxmlc] 
    [mxmlc] /Users/marchbold/work/distriqt/airsdk/tests/mxmlc-anes/src/TestMXMLC.as:4
    [mxmlc] Warning: Definition com.distriqt.extension.core.Core could not be found.
    [mxmlc]     import com.distriqt.extension.core.Core;
    [mxmlc]            ^
    [mxmlc] 
    [mxmlc] /Users/marchbold/work/distriqt/airsdk/tests/mxmlc-anes/src/TestMXMLC.as:10
    [mxmlc] Error: Access of possibly undefined property Core.
    [mxmlc]             Core.init();
    [mxmlc]             ^
    [mxmlc] 

BUILD FAILED
/Users/marchbold/work/distriqt/airsdk/tests/mxmlc-anes/build.xml:13: mxmlc task failed

Total time: 0 seconds
```

**Reply by ajwfrost** [reply]
Ah I see ... so yes, the flexTasks.jar file appears to be doing some initial configuration/checking before loading in the compiler.jar to do the actual work. And yes, there's a library configuration file that is doing a check on just ".swc" as the extension.

Should be easy to update I think...

**Reply by ajwfrost** [reply]
Try with the jar file in here...
[flexTasks.zip](https://github.com/airsdk/Adobe-Runtime-Support/files/13247900/flexTasks.zip)

**Reply by marchbold** [reply]
AMAZING!! 

That did it! Love ya work @ajwfrost !!!

**Reply by ajwfrost** [reply]
Great .. will get that put in to the next release.
thanks
